### PR TITLE
Feature #252 - Show human readable address on balance route.

### LINKED ISSUE #252: Balance Page : Show human readable address, if available
Detect if the wallet user has a human readable name. If not display a reminder notification “Get your human readable address” with a link to the create username page.

The logic for this can be found in the bootSequence in wallet-demo

### PR DESCRIPTION
**Pull request by albertandrejev**
**Description**
This PR will add support of BNS name to the balance route component. It will close #252.

**Without name**
balance-without-name

**With name**
balance-need-funds

### CODE REVIEW

File: packages/bierzo-wallet/src/routes/balance/components/index.tsx [hunk 1]
```diff
@@ -72,7 +72,7 @@ const BalanceLayout = ({ name, tokens, onSendPayment, onReceivePayment }: Props)
       <Block bgcolor={theme.palette.background.paper} height="unset" width={450}>
         <Block padding={4} display="flex" flexDirection="column">
           <Typography variant="h5" align="center" weight="light">
-            {name ? name : '--'}
+            {iovAddress ? iovAddress : 'Get your human readable address.'}
```
Comment (apanizo) [inline packages/bierzo-wallet/src/routes/balance/components/index.tsx]: Can you check with @kmw101 and @webmaster128 if they give thumbs up for the 'Get your human-readable address' text?

File: packages/bierzo-wallet/src/routes/balance/index.dom.spec.ts [hunk 2]
```diff
@@ -24,6 +25,15 @@ const balancesAmount: DeepPartial<BalanceState> = {
   },
 };
 
+const bnsChaingId = 'local-bns-devnet';
```
Comment (apanizo) [inline packages/bierzo-wallet/src/routes/balance/index.dom.spec.ts]: typo, should be -> bnsChainId

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts [hunk 3]
```diff
@@ -55,4 +56,10 @@ withChainsDescribe('E2E > Balance route', (): void => {
     expect(secondBalance).toBe('10 CASH');
     expect(thirdBalance).toBe('10 ETH');
   }, 45000);
+
+  fit('should contain message to get username', async (): Promise<void> => {
```
Comment (apanizo) [inline packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: fit?

File: packages/bierzo-wallet/src/store/usernames/selectors.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { RootState } from '../reducers';
+import { BwUsername } from '.';
+
+export const getUsername = (state: RootState): BwUsername | undefined =>
+  Object.values(state.usernames).find(username => username);
```
Comment (apanizo) [inline packages/bierzo-wallet/src/store/usernames/selectors.ts]: What is the reasoning behind this? Get the first username? Why? Are we assuming they are going to be the same for all addresses?

Can you add a comment with some explanations?

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts (hunk 3 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: Also, we do not have an e2e integration test for showing a real human readable address, right?

Comment (albertandrejev) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: Yes we don't, but how to show it in e2e test? I wanted to ask this too.

File: packages/bierzo-wallet/src/store/usernames/selectors.ts (hunk 4 above)
Comment (albertandrejev) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: 1. This is not possible to show different usernames according to current wallet spec.
2. I saw this code snippet in `wallet-demo`: 
```
const bnsConn = value.connection as BnsConnection; dispatch(setBnsChainId(bnsConn.chainId()));
```
which shows me that `bnsConn.chainId()` contains bns chain id.
And I decided as we are checking for Bns in [/usernames/actions.ts#L17](https://github.com/iov-one/ponferrada/blob/85e6ed75847d4282684e5d7f4b63706166c3f755/packages/bierzo-wallet/src/store/usernames/actions.ts#L17)  and according to the whole of this method logic then I can use first record from `usernames` state as username of the current logged in user.

Correct me if I am wrong. @webmaster128.

File: packages/bierzo-wallet/src/routes/balance/components/index.tsx (hunk 1 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/routes/balance/components/index.tsx]: 👌

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts (hunk 3 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: In this case, let @webmaster128 know we need to have a ticket for having an e2e test where:

1. We signup in the app using the extension
2. We buy a human-readable address
3. We travel to the account balance view
4. We see the human-readable address just bought. 

Also, do the same and assure if we just reload, the page, we do login and the human-readable address is there by default.

File: packages/bierzo-wallet/src/store/usernames/selectors.ts (hunk 4 above)
Comment (webmaster128) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: Logic is okay for now. Multiple usernames are an edge case we can take care of later on.

But please update code a little make explicit this is a "first" getter. By the way, the simpler way to implement "first§ in a type-safe way is `.find(() => true)`.

```ts
export const getUsername = (state: RootState): BwUsername | undefined => {
  const firstUsername = Object.values(state.usernames).find(() => true);
  return firstUsername;
}
  
```

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts (hunk 3 above)
Comment (webmaster128) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: @albertandrejev @apanizo if you want to extract work into a separate ticket, please create that on your own. There is no need for me to think about every single tickets that is created, especially when you are the experts of the particular job.

File: packages/bierzo-wallet/src/store/usernames/selectors.ts (hunk 4 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: > This is not possible to show different usernames according to current wallet spec.

I am not saying you need to support multiple usernames (one by chain), but if we just take the first one, due a product restriction, we should comment it. For avoid confusion.

If you take a look into the usernames reducer you will see an array of usernames, one per key. In the wallet demo we had the same username for all chains, but unless I am wrong we could have one per chain.

My point is, seeing the username is tied to a chain id, having a `getUsername` selector without having specified somehow the chainId is kind of incorrect. I suggest:
```
const getUsernameByChainId = createSelector(
  state => Object.values(state.usernames),
  (_, chainId) => chainId,
  (usernames, chainId) => usernmaes.find(username => username.chainId === chainId)
)

....
// in react component
const username = useSelector(state =>
    getUsernameByChainId(state, 'bnsChainId')
  ))
```
👀 👀 👀 
For now, I suggest to hardcode the bnsChainId, but contact @webmaster128 and ask **him to create a ticket** for introducing in the extension reducer an action for spec the bns chain id as separate property:
```
export interface ExtensionState {
  readonly connected: boolean;
  readonly installed: boolean;
  readonly keys: { [chain: string]: string };
  readonly bnsChainId: string;
}
```
Or if you think this is messing up things, just create `src/store/chains` reducer and place there the chainIds of supported chains.

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts (hunk 3 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: oh! Thanks for the clarification, thought everything had to go through you. Maybe we should involve Martin for validating them?

Comment (webmaster128) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: I assume everything is valid by default. If I find something that makes no sense, we can discuss and invalidate/adjust/close on demand. You guys know what you are doing. I just take care of prioritization.

File: packages/bierzo-wallet/src/store/usernames/selectors.ts (hunk 4 above)
Comment (apanizo) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: @webmaster128 can you validate the reasoning here and then, @albertandrejev create a ticket pitching Martin for validating it?

File: packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts (hunk 3 above)
Comment (albertandrejev) [reply packages/bierzo-wallet/src/routes/balance/index.e2e.spec.ts]: https://github.com/iov-one/ponferrada/issues/405

File: packages/bierzo-wallet/src/store/usernames/selectors.ts (hunk 4 above)
Comment (webmaster128) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: > My point is, seeing the username is tied to a chain id, having a getUsername selector without having specified somehow the chainId is kind of incorrect.

Usernames only exist ion the BNS chain. What we want to show is all usernames that the user owns (typically 0 or 1). I will go though the code that creates this data today and try to clarify the situation. Albert's use makes sense to me.

Comment (albertandrejev) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: Yes, this was my first idea is to hardcode chain id.

Comment (albertandrejev) [reply packages/bierzo-wallet/src/store/usernames/selectors.ts]: @apanizo #408.

**Pull request merged**